File: packages/ts/src/a2a/routing.ts
/**
 * Versioned identifier for the AI37 Agent Card routing extension.
 * It is a namespace, not an endpoint: consumers must not dereference it.
 */
export const AI37_ROUTING_EXTENSION_URI =
  'https://schemas.ai37.ru/a2a/extensions/routing/v1' as const

export const AI37_ROUTING_INTENTS = [
  'document_search',
  'document_list',
  'normative_qa',
  'exact_clause_lookup',
  'engineering_calculation',
  'parameter_selection',
  'counterparty_verification',
  'file_processing',
  'workflow_continue',
] as const

export type AgentRoutingIntent = (typeof AI37_ROUTING_INTENTS)[number]

export interface AgentRoutingProfile extends Record<string, unknown> {
  domains: string[]
  intents: AgentRoutingIntent[]
  excludes: string[]
}

export interface AgentRoutingExtension {
  uri: typeof AI37_ROUTING_EXTENSION_URI
  description: string
  required: false
  params: AgentRoutingProfile
}

const limits = {
  domains: { items: 12, length: 80 },
  intents: { items: 16 },
  excludes: { items: 12, length: 160 },
} as const

const allowedIntents = new Set<string>(AI37_ROUTING_INTENTS)

function compactText(value: string): string {
  return Array.from(value, (character) => {
    const code = character.charCodeAt(0)
    return code < 32 || code === 127 ? ' ' : character
  })
    .join('')
    .replace(/[<>]/g, ' ')
    .replace(/\s+/g, ' ')
    .trim()
}

function normalizeStrings(
  value: unknown,
  field: 'domains' | 'excludes',
): string[] {
  if (!Array.isArray(value)) {
    throw new TypeError(`routing.${field} must be an array`)
  }
  const limit = limits[field]
  if (value.length > limit.items) {
    throw new RangeError(`routing.${field} must contain at most ${limit.items} items`)
  }
  const result: string[] = []
  const seen = new Set<string>()
  for (const item of value) {
    if (typeof item !== 'string') {
      throw new TypeError(`routing.${field} items must be strings`)
    }
    const normalized = compactText(item)
    if (!normalized || normalized.length > limit.length) {
      throw new RangeError(`routing.${field} items must be 1..${limit.length} characters`)
    }
    const key = normalized.toLocaleLowerCase('ru')
    if (!seen.has(key)) {
      seen.add(key)
      result.push(normalized)
    }
  }
  return result
}

function normalizeIntents(value: unknown): AgentRoutingIntent[] {
  if (!Array.isArray(value)) {
    throw new TypeError('routing.intents must be an array')
  }
  if (value.length > limits.intents.items) {
    throw new RangeError(`routing.intents must contain at most ${limits.intents.items} items`)
  }
  const result: AgentRoutingIntent[] = []
  for (const item of value) {
    if (typeof item !== 'string' || !allowedIntents.has(item)) {
      throw new TypeError(`unsupported routing intent: ${String(item)}`)
    }
    if (!result.includes(item as AgentRoutingIntent)) {
      result.push(item as AgentRoutingIntent)
    }
  }
  return result
}

export function normalizeAgentRoutingProfile(value: unknown): AgentRoutingProfile {
  if (!value || typeof value !== 'object' || Array.isArray(value)) {
    throw new TypeError('routing profile must be an object')
  }
  const profile = value as Record<string, unknown>
  return {
    domains: normalizeStrings(profile.domains, 'domains'),
    intents: normalizeIntents(profile.intents),
    excludes: normalizeStrings(profile.excludes, 'excludes'),
  }
}

export function buildAgentRoutingExtension(
  profile: AgentRoutingProfile,
): AgentRoutingExtension {
  return {
    uri: AI37_ROUTING_EXTENSION_URI,
    description: 'Compact semantic routing profile for the AI37 agent registry.',
    required: false,
    params: normalizeAgentRoutingProfile(profile),
  }
}

export function parseAgentRoutingExtension(
  extensions: readonly unknown[] | null | undefined,
): AgentRoutingProfile | undefined {
  const extension = extensions?.find(
    (item) =>
      !!item &&
      typeof item === 'object' &&
      (item as Record<string, unknown>).uri === AI37_ROUTING_EXTENSION_URI,
  ) as Record<string, unknown> | undefined
  if (!extension) return undefined
  try {
    return normalizeAgentRoutingProfile(extension.params)
  } catch {
    return undefined
  }
}

File: packages/python/src/ai37_agent_sdk/a2a/routing.py
from __future__ import annotations

from collections.abc import Iterable, Mapping
from typing import Any, Literal, TypedDict, cast

# Versioned identifier, not an endpoint. Consumers must not dereference it.
AI37_ROUTING_EXTENSION_URI = "https://schemas.ai37.ru/a2a/extensions/routing/v1"

# Operation intents only. HITL/wizard continuation is orchestrator mode, not a card intent.
AgentRoutingIntent = Literal[
    "document_search",
    "document_list",
    "normative_qa",
    "exact_clause_lookup",
    "engineering_calculation",
    "parameter_selection",
    "counterparty_verification",
    "file_processing",
]

AI37_ROUTING_INTENTS: frozenset[str] = frozenset(
    {
        "document_search",
        "document_list",
        "normative_qa",
        "exact_clause_lookup",
        "engineering_calculation",
        "parameter_selection",
        "counterparty_verification",
        "file_processing",
    }
)


class AgentRoutingProfile(TypedDict):
    domains: list[str]
    intents: list[AgentRoutingIntent]
    excludes: list[str]


class AgentRoutingExtension(TypedDict):
    uri: str
    description: str
    required: bool
    params: AgentRoutingProfile


def _compact_text(value: str) -> str:
    safe = "".join(
        " " if ord(character) < 32 or ord(character) == 127 else character for character in value
    )
    return " ".join(safe.replace("<", " ").replace(">", " ").split())


def _strings(value: object, field: str, *, max_items: int, max_length: int) -> list[str]:
    if not isinstance(value, list):
        raise TypeError(f"routing.{field} must be an array")
    if len(value) > max_items:
        raise ValueError(f"routing.{field} must contain at most {max_items} items")
    result: list[str] = []
    seen: set[str] = set()
    for item in value:
        if not isinstance(item, str):
            raise TypeError(f"routing.{field} items must be strings")
        normalized = _compact_text(item)
        if not normalized or len(normalized) > max_length:
            raise ValueError(f"routing.{field} items must be 1..{max_length} characters")
        key = normalized.casefold()
        if key not in seen:
            seen.add(key)
            result.append(normalized)
    return result


def normalize_agent_routing_profile(value: object) -> AgentRoutingProfile:
    if not isinstance(value, Mapping):
        raise TypeError("routing profile must be an object")
    raw_intents = value.get("intents")
    if not isinstance(raw_intents, list):
        raise TypeError("routing.intents must be an array")
    if len(raw_intents) > 16:
        raise ValueError("routing.intents must contain at most 16 items")
    intents: list[AgentRoutingIntent] = []
    for item in raw_intents:
        if not isinstance(item, str) or item not in AI37_ROUTING_INTENTS:
            raise TypeError(f"unsupported routing intent: {item}")
        intent = cast(AgentRoutingIntent, item)
        if intent not in intents:
            intents.append(intent)
    return {
        "domains": _strings(value.get("domains"), "domains", max_items=12, max_length=80),
        "intents": intents,
        "excludes": _strings(value.get("excludes"), "excludes", max_items=12, max_length=160),
    }


def build_agent_routing_extension(profile: AgentRoutingProfile) -> AgentRoutingExtension:
    return {
        "uri": AI37_ROUTING_EXTENSION_URI,
        "description": "Compact semantic routing profile for the AI37 agent registry.",
        "required": False,
        "params": normalize_agent_routing_profile(profile),
    }


def parse_agent_routing_extension(
    extensions: Iterable[object] | None,
) -> AgentRoutingProfile | None:
    for item in extensions or ():
        if not isinstance(item, Mapping) or item.get("uri") != AI37_ROUTING_EXTENSION_URI:
            continue
        try:
            return normalize_agent_routing_profile(cast(Mapping[str, Any], item).get("params"))
        except (TypeError, ValueError):
            return None
    return None
